### ISSUE TITLE
Exclude a package from packagemanager extension list.

### ISSUE THREAD

**Issue by BoPeng** [jupyterlab/jupyterlab#5057]
I had an early version of sos extension that was published on npm as [`sos-extension`](https://www.npmjs.com/package/sos-extension). It was soon renamed to [`jupyterlab-sos`](https://www.npmjs.com/package/jupyterlab-sos) but there was no way to unpublish `sos-extension`. The JupyterLab package manager searches extensions by keyword `jupyterlab-extension` and find both packages.

I tried to publish a newer version of `sos-extension` without `jupyterlab-extension` keyword but npm still returns `sos-extension`, I suppose because the original version had this keyword. Is there anything else I can do to exclude it from the packagemanager? Perhaps JLab can exclude packages with keyword `deprecated`? 

On a broader scale, there will be packages that are left unmaintained, with outdated `dependencies` that do not work with the current version of JLab. It would be useful to grab the full package information and filter by dependencies.

**Comment by blink1073** [comment]
:+1: to ignore `deprecated` packages.  This might be a good reason to introduce an `engines` field as well, similar to what [atom](https://flight-manual.atom.io/hacking-atom/sections/package-word-count/#packagejson) and [vscode](https://code.visualstudio.com/docs/extensionAPI/extension-manifest) use, as a top level screen for compatibility.

**Comment by BoPeng** [comment]
The `engine` information is already in `dependencies`, e.g.

```
 "dependencies": {
    "@jupyterlab/application": "^0.17.2",
    "@jupyterlab/apputils": "^0.17.2",
```
but [`npm v1/search`](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md) does not return such information, and grabbing the details of each package might be slow.

On the other hand, whereas it is useful to check dependencies now, all extensions are supposed to work forever if they use the stable JLab 1.0 api, so there is less a need to check it, right?

**Comment by blink1073** [comment]
Wow, it is surprising that you can't get the `engines` field from their API.  I disagree that grabbing the dependencies would be slow relative to the cost of the overall request and provides valuable information to the user.

**Comment by BoPeng** [comment]
Is there existing mechanism/code to check if an extension is compatible with existing version of JLab? If so, I can try to grab the dependency information of packages and discard incompatible ones.

**Comment by blink1073** [comment]
Yep!  https://github.com/jupyterlab/jupyterlab/blob/d7ceeca96e1dec4c1518a1fc81e08eaf0dc3958d/jupyterlab/commands.py#L355

**Comment by BoPeng** [comment]
OK, does the following sound like a good plan?

1. Extend [extension manager API](https://github.com/jupyterlab/jupyterlab/blob/d7ceeca96e1dec4c1518a1fc81e08eaf0dc3958d/jupyterlab/extension_manager_handler.py#L206) with a command like `validate`.
2. From the frontend, after getting a list of extensions from `npm`, call this API with a list of package names.
3. The API returns a list of compatible extensions for the frontend to display.

**Comment by blink1073** [comment]
Sounds right to me!

**Comment by xiaoyekong** [comment]
我正在做jupyterlab的侧边栏扩展功能，但是不知道该如何去正确创建这个扩展？现在我在能帮我描述一下创建过程么？

**Comment by xiaoyekong** [comment]
我正在做jupyterlab的侧边栏扩展功能，但是源码结构过于复杂，不太容易理解，不知道该如何去正确创建这个扩展？现在我尝试增加了extension文件并修改metapackage下的package.json和index.ts来引入我增添的extension，extension中的代码是执行的，但是执行代码时，页面内容还未加载出来，不能进行dom操作。可以帮我看看是哪里出问题了么？或者我拓展方式是错误的？

**Comment by BoPeng** [comment]
这个版基本没人说中文，你的程序在哪里？github? 我最近要准备JupyterCon，可能没时间看你的程序，不好意思。

**Comment by BoPeng** [comment]
Yes, I was also trying to figure out a good caching mechanism to avoid calling npm for each package for each update and for each instance of JLab started. In addition to client and server side caches, I even think JLab can be distributed with a white list of commonly used packages, even a snapshot of all compatible extensions at the time of each JLab release. Dumping such a list as part of the release script should not be terribly difficult and there would be less a need for complex caching mechanisms if JLab only needs to check compatibility of new extensions.
